### PR TITLE
Test | Seperate LocalDB and NamedPipe LocalDB Tests

### PR DESCRIPTION
**Pull request by Kaur-Parminder**
- Seperated LocalDb and Named Pipe tests, added configuration for same.
- Added a fix for exceuteLocalDbCommandProcess which throws array out of bound when running "sqllocaldb start localdbname"
- Added LocalDB restart for some methods that intermittently fail on pipeline in case LocalDb stops
Run can be seen here https://dev.azure.com/SqlClientDrivers/public/_build/results?buildId=45081&view=results [There are file stream failures, #1458  PR address those test failure]

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 1]
```diff
@@ -90,6 +89,34 @@ public static void SqlLocalDbSharedInstanceConnectionTest()
         }
         #endregion
 
+        #region NamedPipeTests
+
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)] // No Registry support on UAP
+        [ConditionalFact(nameof(IsLocalDbNamedPipeSet))]
+        public static void SqlLocalDbNamedPipeConnectionTest()
+        {
+            ConnectionTest(s_localDbConnectionString);
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:98]: Shouldn't this test only be testing localdb named pipes?

File: src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/config.default.json [hunk 2]
```diff
@@ -17,7 +17,8 @@
     "AzureKeyVaultClientSecret": "",
     "SupportsIntegratedSecurity": true,
     "LocalDbAppName": "",
-    "LocalDbSharedInstanceName":"",
+    "LocalDbSharedInstanceName": "",
+    "LocalDbNamedPipeSet": false,
```
Comment (johnnypham) [inline src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/config.default.json:21]: Let's be consistent with the other keys and use either `LocalDbNamedPipesEnabled` or `SupportsLocalDbNamedPipes`.

**Review by johnnypham — APPROVED** [review]

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 3]
```diff
@@ -133,16 +159,20 @@ private static void OpenConnection(string connString)
         }
 
         private static string GetLocalDbNamedPipe()
+        {
+            RestartLocalDB();
+            return ExecuteLocalDBCommandProcess(s_commandPrompt, s_sqlLocalDbInfo, "pipeName");
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:164]: Could you declare constants for "pipeName" and "state" or define an Enum?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 4]
```diff
@@ -155,6 +185,12 @@ private static string ExecuteLocalDBCommandProcess(string filename, string argum
                 RedirectStandardError = true,
             };
             string[] lines = Process.Start(sInfo).StandardOutput.ReadToEnd().Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:187]: Please verify the expected result by running the command. Like assessing the number of returned lines.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 5]
```diff
@@ -155,6 +185,12 @@ private static string ExecuteLocalDBCommandProcess(string filename, string argum
...
+
+            if (arguments == s_startLocalDbCommand)
+            {
+                sInfo.Arguments = s_sqlLocalDbInfo; //after start check info again
+                lines = Process.Start(sInfo).StandardOutput.ReadToEnd().Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:192]: Assess the expected result here too.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs (hunk 3 above)
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:164]: Please, assess the result before returning.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 6]
```diff
@@ -90,6 +89,33 @@ public static void SqlLocalDbSharedInstanceConnectionTest()
         }
         #endregion
 
+        #region NamedPipeTests
+
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)] // No Registry support on UAP
+        [ConditionalFact(nameof(IsLocalDbNamedPipesEnabled))]
+        public static void SqlLocalDbNamedPipeConnectionTest()
+        {
+            ConnectionTest(s_localDbNamedPipeConnectionString);
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:98]: In these tests, you didn't do anything with the instance name while it's checked to be skipped by `IsLocalDbNamedPipesEnabled`!
Did you miss the test or just assess the wrong parameter?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 7]
```diff
@@ -133,16 +159,20 @@ private static void OpenConnection(string connString)
...
         private static string GetLocalDbNamedPipe()
+        {
+            RestartLocalDB();
+            return ExecuteLocalDBCommandProcess(s_commandPrompt, s_sqlLocalDbInfo, "pipeName");
+        }
+
+        private static void RestartLocalDB()
         {
             string state = ExecuteLocalDBCommandProcess(s_commandPrompt, s_sqlLocalDbInfo, "state");
             while (state.Equals("stopped", StringComparison.InvariantCultureIgnoreCase))
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:170]: If the service couldn't start it will be an infinitive loop. Try for 5 times and throw an exception if exceeded.

**Review by DavoudEshtehari — COMMENTED** [review]
There is a table in [BUILDGUIDE.md](https://github.com/dotnet/SqlClient/blob/main/BUILDGUIDE.md) to introduce the test configuration properties. Could you update the table and include the missing properties in addition to this new property?

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs (hunk 6 above)
Comment (Kaur-Parminder) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:103]: Test is using NamedPipe connection string, that is why the annotation is being used.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs (hunk 4 above)
Comment (Kaur-Parminder) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:197]: Added asserts in if/else as they can vary

**Comment by Kaur-Parminder** [comment]
> There is a table in [BUILDGUIDE.md](https://github.com/dotnet/SqlClient/blob/main/BUILDGUIDE.md) to introduce the test configuration properties. Could you update the table and include the missing properties in addition to this new property?

Added.

**Comment by DavoudEshtehari** [comment]
> > There is a table in [BUILDGUIDE.md](https://github.com/dotnet/SqlClient/blob/main/BUILDGUIDE.md) to introduce the test configuration properties. Could you update the table and include the missing properties in addition to this new property?
> 
> Added.

Please, add all the missed properties. You can compare this table with [config.default.json](https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/config.default.json) to find the differences.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 8]
```diff
@@ -134,16 +165,25 @@ private static void OpenConnection(string connString)
...
+        {
+            string state = ExecuteLocalDBCommandProcess(s_commandPrompt, s_sqlLocalDbInfo, InfoType.state);
+            int count = 5;
+            while (state.Equals("stopped", StringComparison.InvariantCultureIgnoreCase) && count>0)
             {
-                state = ExecuteLocalDBCommandProcess(s_commandPrompt, s_startLocalDbCommand, "state");
+                count--;
+                state = ExecuteLocalDBCommandProcess(s_commandPrompt, s_startLocalDbCommand, InfoType.state);
                 Thread.Sleep(2000);
             }
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:184]: Regarding this [comment](https://github.com/dotnet/SqlClient/pull/1499#discussion_r798990435), you missed throwing the exception.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 9]
```diff
@@ -155,12 +195,22 @@ private static string ExecuteLocalDBCommandProcess(string filename, string argum
...
             string[] lines = Process.Start(sInfo).StandardOutput.ReadToEnd().Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
-            if (infoType.Equals("state"))
+
+            if (arguments == s_startLocalDbCommand)
+            {
+                Assert.Equal(2, lines.Length);
+                sInfo.Arguments = s_sqlLocalDbInfo; //after start check info again
+                lines = Process.Start(sInfo).StandardOutput.ReadToEnd().Split(new string[] { Environment.NewLine }, StringSplitOptions.None);
+                Assert.Equal(9, lines.Length);
+            }
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:205]:
```suggestion
            }
            Assert.Equal(9, lines.Length);
```

File: BUILDGUIDE.md [hunk 10]
```diff
@@ -173,6 +173,7 @@ Manual Tests require the below setup to run:
   |AzureKeyVaultClientSecret | (Optional) "Client Secret" of the Active Directory registered application, granted access to the Azure Key Vault specified in `AZURE_KEY_VAULT_URL` | _{Client Application Secret}_ |
   |LocalDbAppName | (Optional) If Local Db Testing is supported, this property configures the name of Local DB App instance available in client environment. Empty string value disables Local Db testing. | Name of Local Db App to connect to.|
   |LocalDbSharedInstanceName | (Optional) If LocalDB testing is supported and the instance is shared, this property configures the name of the shared instance of LocalDB to connect to. | Name of shared instance of LocalDB. |
+  |LocalDbNamedPipesEnabled | (Optional) If Local Db Testing is supported, this property enables the run of Local db named pipe instance tests, LocalDbAppName should also ber set | `true` OR `false`|
```
Comment (DavoudEshtehari) [inline BUILDGUIDE.md:176]: This property is redundant. When a Localdb is running, the **Instance pipe name** will be available.

Comment (Kaur-Parminder) [reply BUILDGUIDE.md]: I remove this configuration and added the Active issue as discussed.

**Comment by Kaur-Parminder** [comment]
> > > There is a table in [BUILDGUIDE.md](https://github.com/dotnet/SqlClient/blob/main/BUILDGUIDE.md) to introduce the test configuration properties. Could you update the table and include the missing properties in addition to this new property?
> > 
> > 
> > Added.
> 
> Please, add all the missed properties. You can compare this table with [config.default.json](https://github.com/dotnet/SqlClient/blob/main/src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/config.default.json) to find the differences.

I added few and skipped the ones, I was not sure about.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 11]
```diff
@@ -90,6 +94,36 @@ public static void SqlLocalDbSharedInstanceConnectionTest()
         }
         #endregion
 
+        #region NamedPipeTests
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)] // No Registry support on UAP
+        [ActiveIssue(20244)] //pending pipeline configuration
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:101]:
```suggestion
        [ActiveIssue(20245)] //pending pipeline configuration
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 12]
```diff
@@ -90,6 +94,36 @@ public static void SqlLocalDbSharedInstanceConnectionTest()
...
+        public static void SqlLocalDbNamedPipeConnectionTest()
+        {
+            ConnectionTest(s_localDbNamedPipeConnectionString);
+        }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)] // No Registry support on UAP
+        [ActiveIssue(20244)] //pending pipeline configuration
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:109]:
```suggestion
        [ActiveIssue(20245)] //pending pipeline configuration
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs [hunk 13]
```diff
@@ -90,6 +94,36 @@ public static void SqlLocalDbSharedInstanceConnectionTest()
...
+        public static void LocalDBNamedPipeEncryptionNotSupportedTest()
+        {
+            // Encryption is not supported by SQL Local DB.
+            // But connection should succeed as encryption is disabled by driver.
+            ConnectionWithEncryptionTest(s_localDbNamedPipeConnectionString);
+        }
+
+        [Fact]
+        [SkipOnTargetFramework(TargetFrameworkMonikers.Uap)] // No Registry support on UAP
+        [ActiveIssue(20244)] //pending pipeline configuration
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/SQL/LocalDBTest/LocalDBTest.cs:119]:
```suggestion
        [ActiveIssue(20245)] //pending pipeline configuration
```

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs [hunk 14]
```diff
@@ -466,7 +466,6 @@ public static void DropDatabase(SqlConnection sqlConnection, string dbName)
 
         public static bool IsLocalDBInstalled() => !string.IsNullOrEmpty(LocalDbAppName?.Trim()) && IsIntegratedSecuritySetup();
         public static bool IsLocalDbSharedInstanceSetup() => !string.IsNullOrEmpty(LocalDbSharedInstanceName?.Trim()) && IsIntegratedSecuritySetup();
-
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs:469]: Unnecessary change! Could be reverted.

File: src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/config.default.json [hunk 15]
```diff
@@ -17,7 +17,7 @@
     "AzureKeyVaultClientSecret": "",
     "SupportsIntegratedSecurity": true,
     "LocalDbAppName": "",
-    "LocalDbSharedInstanceName":"",
+    "LocalDbSharedInstanceName": "",
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/tools/Microsoft.Data.SqlClient.TestUtilities/config.default.json:20]: Unnecessary change! Could be reverted.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs (hunk 14 above)
Comment (DavoudEshtehari) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs:469]: still not reverted!

Comment (Kaur-Parminder) [reply src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs:469]: Weird, Can you make new line suggestion please, not sure by new line did not mark it outdated.

File: src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs [hunk 16]
```diff
@@ -466,7 +466,7 @@ public static void DropDatabase(SqlConnection sqlConnection, string dbName)
...
+
```
Comment (DavoudEshtehari) [inline src/Microsoft.Data.SqlClient/tests/ManualTests/DataCommon/DataTestUtility.cs:469]:
```suggestion
```

**Review by DavoudEshtehari — APPROVED** [review]

**Review by JRahnama — APPROVED** [review]

**Review by DavoudEshtehari — APPROVED** [review]
